src/Client/Speedrun.java

```diff
@@ -0,0 +1,329 @@
+        name, ticks, calcTimeSinceLastSegment(time)), Client.CHAT_QUEST);
+  }
+
+  public static void printTimeSinceLastSegment(boolean referenceEndTimes) {
+    long timeDelta;
+    if (referenceEndTimes) {
+      timeDelta = endTimes.get(endTimes.size() - 1) - startTimes.get(startTimes.size() - 1);
+    } else {
+      timeDelta = calcTimeSinceLastSegment(System.currentTimeMillis());
+    }
+    Client.displayMessage(String.format("Millis since last time segment: %d", timeDelta), Client.CHAT_QUEST);
+  }
+
+  public static long calcTimeSinceLastSegment(long millis) {
+    int startSize = startTimes.size();
+    if (startSize >= 1)
+      return millis - startTimes.get(startSize - 1);
+    else
+      return millis;
+  }
+
+  public static void incrementTicks() {
+    if (active)
+      totalTicks++;
+  }
+
+  public static void printTotalTicks() {
+    Client.displayMessage(String.format("Total ticks spent in speedrun: %d", totalTicks), Client.CHAT_QUEST);
+  }
+
+  public static void saveAndQuitSpeedrun() {
+    loadResult = REASON_NEED_LOAD;
+    if (active) {
+      active = false;
+      addEndTime(System.currentTimeMillis());
+      try {
+        DataOutputStream speedrunData = new DataOutputStream(
+          new BufferedOutputStream(
+              new FileOutputStream(
+                  new File(
+                  Settings.Dir.SPEEDRUN + String.format(
+                      "/%d.%s.bin", startTimes.get(0), Client.username_login.replaceAll("[^a-zA-Z0-9]", ""))
+                  )
+              )
+          )
+        );
+
+        // Header information
+        speedrunData.writeInt(1533546026); // unix timestamp that RSC was taken offline
+        speedrunData.writeBoolean(!endTheRUNNN);
+        if (endTheRUNNN) {
+          endTheRUNNN = false;
+        }
+        speedrunData.writeInt(totalTicks);
+        speedrunData.writeInt(startTimes.size());
+        speedrunData.writeInt(endTimes.size());
+        speedrunData.writeInt(goalsDefined);
+        speedrunData.writeInt(Client.username_login.length());
+        for (int i = 0; i < 32; i++)
+          speedrunData.writeInt(0); // reserved
+
+        // Speedrun data
+        for (long time : startTimes) {
+          speedrunData.writeLong(time);
+        }
+        for (long time : endTimes) {
+          speedrunData.writeLong(time);
+        }
+        for (int i = 0; i < goalsDefined; i++) {
+          speedrunData.writeInt(completionTicks[i]);
+          speedrunData.writeLong(completionTimes[i]);
+        }
+        speedrunData.writeChars(Client.username_login);
+
+        speedrunData.flush();
+        speedrunData.close();
+      } catch (IOException e) {
+        Logger.Error("@|red Couldn't write speedrun save file!|@");
+        Client.displayMessage("@red@Couldn't write speedrun save file!", Client.CHAT_QUEST);
+        Client.displayMessage("Here are your times:", Client.CHAT_QUEST); // TODO
```

Comment:
the todo will be placed in this PR or left for another occasion?

Reply:
I forgot that I put that, but actually yea, I'm planning on adding a keybind later to print out current times, so, it could be part of that.

Reply:
sounds good